ntbs-service/DataMigration/ImportValidator.cs

```diff
@@ -95,6 +99,44 @@ public ImportValidator(IImportLogger logger, IReferenceDataRepository referenceD
             }
         }
 
+        private static void CleanContainsNoTabsProperties(object entity)
+        {
+            if (entity == null)
+            {
+                return;
+            }
+
+            // Clean up any [ContainsNoTabs] properties on this object
+            entity
+                .GetType()
+                .GetProperties()
+                .Where(prop => Attribute.IsDefined(prop, typeof(ContainsNoTabsAttribute)))
+                .ForEach(prop => prop.SetValue(entity, TrimAndCleanString((string)prop.GetValue(entity))));
+
+            // Clean inside each field
+            entity
+                .GetType()
+                .GetProperties()
+                .Where(prop => prop.PropertyType.Namespace == typeof (Notification).Namespace)
+                .ForEach(prop => CleanContainsNoTabsProperties(prop.GetValue(entity)));
+
+            // Clean inside any enumerable properties
+            entity
+                // Get all non-null enumerable properties
+                .GetType()
+                .GetProperties()
+                .Where(prop => typeof(IEnumerable).IsAssignableFrom(prop.PropertyType))
+                .Select(prop => prop.GetValue(entity))
+                .Where(enumerable => enumerable != null)
+                .OfType<IEnumerable<object>>()
+                // Flatten
+                .SelectMany(element => element)
+                .Where(element => element.GetType().Namespace == typeof (Notification).Namespace)
+                .ForEach(CleanContainsNoTabsProperties);
```

Comment:
I wanted to have a go with reflection, with this being the first time I've properly used it, though it has turned out to a bit longer and more complex than I hoped. Sorry it's so late given the release as well. What do you think of it? 

The other option is using a much simpler function, just of lines 110-114 roughly, and a block along the lines of:
```Csharp
CleanContainsNoTabsProperties(notification.ClinicalDetails);
notification.SocialContextAddresses.ForEach(CleanContainsNoTabsProperties);
notification.SocialContextVenues.ForEach(CleanContainsNoTabsProperties);
<etc>
```

Which do you prefer @JoshCadney?

Reply:
I think that for now we should go for the latter approach, for a couple of reasons:

1. Lines 110-114 are entirely general (i.e. you could apply them to any object and they'd behave how you expect) but the subsequent lines have a dependency on the Notification namespace, which introduces a coupling between this method and how it's being used (i.e. to validate notifications).
2. There's a risk of a stack overflow - if we applied this method to an object with a circular recursion (which is pretty common in EF). In fact, I think a Notification already has some circular references, so the success of this code depends on how the object is populated. Now, we could add a max depth, or something more clever here, but I think it's better to keep it simple.